### PR TITLE
feat: Plain integration

### PR DESCRIPTION
**Pull request by nizzyabi**
## What does this PR do?

This is a Cal & Plain.com integration that allows us to display customer cards on plain.com to speed up the process of helping them.

- Fixes #XXXX (GitHub issue number)
- Fixes CAL-XXXX (Linear issue number - should be visible at the bottom of the GitHub issue description)

will send one privately, includes env file

## Mandatory Tasks (DO NOT REMOVE)

- [y] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [y] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [y] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

- Are there environment variables that should be set?
         yes: PLAIN_API_KEY, PLAIN_API_URL, HMAC_SECRET_KEY
- What are the minimal test data to have? 
         users in cal.com local environment
- What is expected (happy path) to have (input and output)? 
        you can see the user card in plain.com
- Any other important info that could help to test that PR

## Checklist

- I haven't read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)
- My code doesn't follow the style guidelines of this project
- I haven't commented my code, particularly in hard-to-understand areas
- I haven't checked if my changes generate no new warnings

### CODE REVIEW

File: packages/app-store/office365calendar/api/add.ts
Comment (baileypumfleet) [inline packages/app-store/office365calendar/api/add.ts:1]: So you know why this and the yarn.lock are in the PR?

File: apps/web/pages/api/get-customer.ts [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+  componentText?: {
+    text: string;
+  };
+  componentDivider?: {
+    dividerSpacing?: "relaxed" | "default" | "compact";
+  };
+  componentHeader?: {
+    text: string;
+  };
+};
```
Comment (youneshenniwrites) [inline apps/web/pages/api/get-customer.ts:32]:
```suggestion
enum DividerSpacing {
  Relaxed = "relaxed",
  Default = "default",
  Compact = "compact",
}

type CardComponent = {
  componentText?: {
    text: string;
  };
  componentDivider?: {
    dividerSpacing?: DividerSpacing;
  };
  componentHeader?: {
    text: string;
  };
};
```

File: apps/web/pages/api/example.ts [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+    const { email, id, username, timeZone, emailVerified, plan, identityProvider, twoFactorEnabled } =
+      customer;
+
+    if (!email) {
+      return res.status(400).json({ error: "Invalid customer data" });
+    }
+
+    // Fetch user details from the database using Prisma
+    const user = await prisma.user.findUnique({
+      where: { email },
```
Comment (baileypumfleet) [inline apps/web/pages/api/example.ts:35]: Let's only fetch certain fields based on what we need. The user table has a lot of columns nowadays

File: apps/web/pages/api/get-customer.ts [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+export default async function handler(req: NextApiRequest, res: NextApiResponse) {
+  if (req.method !== "POST") {
+    res.setHeader("Allow", ["POST"]);
+    return res.status(405).end(`Method ${req.method} Not Allowed`);
+  }
+
+  try {
+    console.log("Request received:", {
```
Comment (baileypumfleet) [inline apps/web/pages/api/get-customer.ts:51]: We should probably get rid of all console.logs that don't log an actual error or important action

File: apps/web/pages/api/plain/customer-card.ts
Comment (baileypumfleet) [inline apps/web/pages/api/plain/customer-card.ts:1]: Is this file still being used? It's all commented out

File: turbo.json
Comment (baileypumfleet) [inline turbo.json:1]: We should also add to .env.example

File: packages/app-store/office365calendar/api/add.ts
Comment (baileypumfleet) [inline packages/app-store/office365calendar/api/add.ts:1]: These two Office365 scope changes still seem to be in this PR

**Review by pumfleet — CHANGES_REQUESTED** [review]
A few comments to address, but looks good mostly :)

File: .env.example [hunk 4]
```diff
@@ -107,6 +107,12 @@ NEXT_PUBLIC_POSTHOG_KEY=
 
 NEXT_PUBLIC_POSTHOG_HOST=
 
+# plain.com config
+
+PLAIN_API_KEY=
+PLAIN_API_URL=https://api.plain.com/v1
+HMAC_SECRET_KEY=
```
Comment (emrysal) [inline .env.example:114]: Can we add a PLAIN_ prefix here too?

File: apps/web/pages/api/customer-card.ts [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+
+    const { cardKeys, customer } = req.body;
+    const { email, id, username, timeZone, emailVerified, identityProvider, twoFactorEnabled } = customer;
+
+    if (!email) {
+      return res.status(400).json({ error: "Invalid customer data" });
+    }
+
+    // Fetch specific user details from the database using Prisma
+    const user = await prisma.user.findUnique({
```
Comment (emrysal) [inline apps/web/pages/api/customer-card.ts:33]: It's better to use the repository/user here 🙏

File: apps/web/pages/api/customer-card.ts [hunk 6]
```diff
@@ -0,0 +1,251 @@
+/* eslint-disable @typescript-eslint/no-non-null-assertion */
+import { cardExamples } from "@pages/api/plain/example-cards";
+import crypto from "crypto";
+import type { NextApiRequest, NextApiResponse } from "next";
+
+import { prisma } from "@calcom/prisma";
+
+export default async function handler(req: NextApiRequest, res: NextApiResponse) {
+  try {
```
Comment (emrysal) [inline apps/web/pages/api/customer-card.ts:9]: Try/catch here is not needed, handlers should get wrapped properly already by the apiHandler

File: apps/web/pages/api/customer-card.ts [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+
+    if (incomingSignature !== expectedSignature) {
+      return res.status(403).send("Forbidden");
+    }
+
+    // Disabled vercel's edge cache
+    res.setHeader("Cache-Control", "no-cache");
+
+    const { cardKeys, customer } = req.body;
+    const { email, id, username, timeZone, emailVerified, identityProvider, twoFactorEnabled } = customer;
```
Comment (emrysal) [inline apps/web/pages/api/customer-card.ts:26]: How is this input validated? Seems to me there's a lot of ways this can 500

**Review by emrysal — CHANGES_REQUESTED** [review]
Needs a bit more work, mostly on the validation & error handling side to be mergeable.

File: packages/lib/CalEventParser.ts [hunk 8]
```diff
@@ -71,42 +40,78 @@ export const getWho = (
...
+  if (calEvent.seatsPerTimeSlot && !calEvent.seatsShowAttendees) {
+    attendeesFromCalEvent = [];
+  }
+  const attendees = attendeesFromCalEvent
+    .map((attendee) => {
+      return `
+${attendee?.name || t("guest")}
+${!isSmsCalEmail(attendee.email) ? `${attendee.email}\n` : `${attendee.phoneNumber}\n`}
+
+`;
```
[the remark this answers is not in the document]
Comment (emrysal) [reply packages/lib/CalEventParser.ts]: This was a revert that somehow showed up as the changelog

**Review by emrysal — APPROVED** [review]
Looks good now, LGTM 🚀

**Review by pumfleet — APPROVED** [review]

**Pull request merged**